dev-packages/cli/src/check-dependencies.ts

```diff
@@ -0,0 +1,306 @@
+    dependent: string | undefined
+}
+
+/** Issue found with a specific package. */
+interface DependencyIssue {
+    /** Type of the issue. */
+    issueType: 'not-hoisted' | 'multiple-versions' | 'theia-version-mix',
+    /** Package with issue. */
+    package: Package,
+    /** Packages related to this issue. */
+    relatedPackages: Package[],
+    /** Severity */
+    severity: 'warning' | 'error'
+}
+
+export default function checkDependencies(options: CheckDependenciesOptions): void {
+    const workspaces = deriveWorkspaces(options);
+    logUpdate(`✅ Found ${workspaces.length} workspaces`);
+
+    console.log('🔍 Collecting dependencies...');
+    const dependencies = findAllDependencies(workspaces, options);
+    logUpdate(`✅ Found ${dependencies.length} dependencies`);
+
+    console.log('🔍 Analyzing dependencies...');
+    const issues = analyzeDependencies(dependencies, options);
+    if (issues.length <= 0) {
+        logUpdate('✅ No issues were found');
+        process.exit(0);
+    }
+
+    logUpdate('🟠 Found ' + issues.length + ' issues');
+    printIssues(issues);
+    printHints(issues);
+    process.exit(options.suppress ? 0 : 1);
+}
+
+function deriveWorkspaces(options: CheckDependenciesOptions): string[] {
+    const wsGlobs = options.workspaces ?? readWorkspaceGlobsFromPackageJson();
+    const workspaces: string[] = [];
+    for (const wsGlob of wsGlobs) {
+        workspaces.push(...glob.sync(wsGlob + '/'));
+    }
+    return workspaces;
+}
+
+function readWorkspaceGlobsFromPackageJson(): string[] {
+    const rootPackageJson = path.join(process.cwd(), PACKAGE_JSON);
+    if (!fs.existsSync(rootPackageJson)) {
+        console.error('Directory does not contain a package.json with defined workspaces');
+        console.info('Run in the root of a Theia project or specify them via --workspaces');
+        process.exit(1);
+    }
+    return require(rootPackageJson).workspaces;
+}
+
+function findAllDependencies(workspaces: string[], options: CheckDependenciesOptions): Package[] {
+    const dependencies: Package[] = [];
+    dependencies.push(...findDependencies('.', options));
+    for (const workspace of workspaces) {
+        dependencies.push(...findDependencies(workspace, options));
+    }
+    return dependencies;
+}
+
+function findDependencies(workspace: string, options: CheckDependenciesOptions): Package[] {
+    const dependent = getPackageName(path.join(process.cwd(), workspace, PACKAGE_JSON));
+    const nodeModulesDir = path.join(workspace, NODE_MODULES);
+    const matchingPackageJsons: string[] = [];
+    for (const include of options.include) {
+        matchingPackageJsons.push(
+            ...glob.sync(include + '/' + PACKAGE_JSON, {
+                cwd: nodeModulesDir,
+                ignore: [`**/${NODE_MODULES}/**`, ...options.exclude]
+            })
+        );
+    }
+    return matchingPackageJsons.map(packageJson => toDependency(packageJson, nodeModulesDir, dependent));
+}
```

Comment:
Please consider the following code that avoids iterating twice over everything:

```suggestion
function findDependencies(workspace: string, options: CheckDependenciesOptions): Package[] {
    const dependent = getPackageName(path.join(process.cwd(), workspace, PACKAGE_JSON));
    const nodeModulesDir = path.join(workspace, NODE_MODULES);
    const matchingPackageJsons: Package[] = [];
    options.include.forEach(include => {
        glob.sync(`${include}/${PACKAGE_JSON}`, {
            cwd: nodeModulesDir,
            ignore: [`**/${NODE_MODULES}/**`, ...options.exclude]
        })
        .forEach(packageJson => {
            matchingPackageJsons.push(toDependency(packageJson, nodeModulesDir, dependent));
        });
    });
    return matchingPackageJsons;
}
```

Reply:
Sure, thanks, that is probably more efficient!